### ISSUE TITLE
AlignHash with separator causing problems

### ISSUE THREAD

**Issue by guyisra** [bbatsov/rubocop#2582]
beyond getting this
```
An error occurred while Style/AlignHash cop was inspecting <file name>
Errors are usually caused by RuboCop bugs.
Please, report your problems to RuboCop's issue tracker.
Mention the following information in the issue report:
0.35.1 (using Parser 2.2.3.0, running on ruby 2.1.2 x86_64-darwin13.0)
```

it also sometimes messes the hash up, the key names etc.

specifically this happened in a code that looks like this, in a spec file

```
let(:params) do
    {
      key1:                    val,
      long_key: looooooong_val,
      some_parm:              {
        key_a: val_a,
        key:       value
      }
    }
  end
```
with the settings
```
Style/AlignHash:
  EnforcedHashRocketStyle: table
  EnforcedColonStyle: seperator
```

**Comment by alexdowad** [comment]
That should be `separator`, not `seperator`. Please try with the correct spelling in the config file.

**Comment by guyisra** [comment]
yea, that was a typo when i created the issue, I typed it by hand.

It still happens, I'll try to create a demo hash where it happens

**Comment by alexdowad** [comment]
Tried the sample code and config which you show above. It doesn't produce an error (at least on `master`).

**Comment by guyisra** [comment]
try this sample, which causes a mashed-up hash as a result and also causing the file to have syntax errors

```
let(:params) do
    {
        sjtjo: sjtjo,
        too_ono_ilitjion_tofotono_o: too_ono_ilitjion_tofotono_o,
        o_lionj_tofotono_o: o_lionj_tofotono_o,
        gtjowty_tofotono_o: gtjowty_tofotono_o,
        o_onnoo_jot_tofotono_o: o_onnoo_jot_tofotono_o,
        o_utono_y: "QWE",
        tmoj: 123,
        o_onnoo_jot: "pt55yon",
        jimosjtmp: jimosjtmp,
        soson_o_lionj_dtjt: soson_o_lionj_dtjt.jo_json,
        o_onnoo_jot_o_lionj_dtjt: { o_onjto_j_id: 'Pt4yon-Vtlijot-o_ttd' }.jo_json,
        o_todij_o_ttd: {
            o_ttd_nbot: o_ttd_nbot,
            btnd: btnd,
            ox_ott: ox_ott,
            op_onjh: op_onjh,
            holot: holot,
            joaon: joaon,
            jypo: o_ttd_jypo
        }
    }
  end
```

**Comment by alexdowad** [comment]
OK, one discovery: there is no problem when auto-correct is not used.

Autocorrection is breaking syntax, causing a syntax error.

**Comment by alexdowad** [comment]
Autocorrection is also eating up some of the key/value pairs. They are completely gone.

**Comment by alexdowad** [comment]
The problem still occurs if the `EnforcedHashRocketStyle` parameter is removed. It appears that it should be related to `EnforcedColonStyle: separator`.

**Comment by alexdowad** [comment]
OK, the problem is happening on line 245 of `align_hash.rb`. If a key/value/etc is too far to the right, the cop deletes some characters on its left side. But it is deleting so many characters, that it takes out part of the token on its left.

**Comment by alexdowad** [comment]
OK, it's getting clearer.

Do you see the key which is `soson_o_lionj_dtjt`? It wants to move that one to the left, so its colon will align with the colon on the previous line.

But this is impossible. The key is so long, and the colon is so much further to the right than the above one, that even if all the whitespace on its left was deleted, it still wouldn't align. But in an attempt to align, the cop is deleting MORE than all the whitespace on its left.

**Comment by alexdowad** [comment]
I have a fix for this, will push soon.

**Comment by guyisra** [comment]
hmmm. I thought it will take the rightest possible colon without deleting and aligning other keys by it

**Comment by alexdowad** [comment]
Pushed a fix to my open PR. The cop could still align the code better, but at least it doesn't break it now.
